### PR TITLE
Migrate to Terraform Plugin Framework

### PR DESCRIPTION
**Pull request by SBGoods**
Closes:
#81 
#80
#77
#173

### CODE REVIEW

File: internal/provider/data_source.go [hunk 1]
```diff
@@ -31,105 +48,111 @@ func dataSource() *schema.Resource {
...
 
-	programI := d.Get("program").([]interface{})
-	workingDir := d.Get("working_dir").(string)
-	query := d.Get("query").(map[string]interface{})
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var program []string
```
Comment (bflad) [inline internal/provider/data_source.go:99]: If there's not already an acceptance test verifying `null` values in the `program` list, it might be good to add one. e.g.

```terraform
data "external" "test" {
  program = ["...", null, "..."]
}
```

I think we may want to err on the side of caution here and use `[]types.String` here, then filter out elements that match `prgramArgRaw.IsNull() || programArgRaw.ValueString() == ""`

We could potentially consider adding explicit validation in the schema to prevent null elements in a future major version.

File: internal/provider/data_source.go [hunk 2]
```diff
@@ -31,105 +48,111 @@ func dataSource() *schema.Resource {
...
+	if len(filteredProgram) == 0 {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("program"),
+			"External Program Missing",
+			"The data source was configured without a program to execute. Verify the configuration contains at least one non-empty value.",
+		)
+		return
+	}
+
+	var query map[string]string
```
Comment (bflad) [inline internal/provider/data_source.go:126]: Similarly with `query`, e.g.

```terraform
data "external" "test" {
  query = {
    test = null
  }
}
```

And using `map[string]types.String` 👍 

We could potentially consider adding explicit validation in the schema to prevent null elements in a future major version or explicitly stating they will always be converted to an empty string.

File: internal/provider/data_source.go [hunk 3]
```diff
@@ -160,66 +184,67 @@ The program must also be executable according to the platform where Terraform is
...
-		}
+`+
+				fmt.Sprintf("\nProgram: %s", cmd.Path)+
+				fmt.Sprintf("\nResult Error: %s", err),
+		)
+		return
 	}
 
-	d.Set("result", result)
+	config.Result, _ = types.MapValueFrom(ctx, types.StringType, result)
```
Comment (bflad) [inline internal/provider/data_source.go:237]: While `map[string]string` should always convert okay given the current code, we should return and handle diagnostics here for defensive programming should the other code change. 👍

File: docs/data-sources/external.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
+---
```
Comment (bflad) [inline docs/data-sources/external.md:1]: Looks like `website/docs` needs to be removed as part of these changes. 😄

**Review by bflad — COMMENTED** [review]
A few minor things, otherwise this is looking great. 👍

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -12,64 +13,76 @@ on:
...
-    - name: TF acceptance tests
-      timeout-minutes: 10
-      env:
-        TF_ACC: "1"
-      run: go test -v -cover ./internal/provider/
+
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # v3.5.0
+        with:
+          go-version-file: 'go.mod'
```
Comment (bendbennett) [inline .github/workflows/test.yml:78]: Looks like the tests are running with Go version as specified in `go.mod` whereas `build` phase is using Go version specified in `strategy: matrix`. Just wanted to check that this is intentional?

**Review by bendbennett — APPROVED** [review]
LGTM 👍

File: .github/workflows/test.yml (hunk 5 above)
Comment (SBGoods) [reply .github/workflows/test.yml:78]: Good callout! I don't think we need the matrix anymore since we are only testing against `1.18` anyways.

File: go.mod [hunk 6]
```diff
@@ -3,20 +3,30 @@ module github.com/terraform-providers/terraform-provider-external
 go 1.18
 
 require (
-	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
+	github.com/hashicorp/terraform-plugin-docs v0.13.0
```
Comment (austinvalle) [inline go.mod:6]: **drive-by reminder:** we should avoid this dependency from upgrading to `v0.14.0` as it'll update `terraform-exec` to `v0.18.0` which will break any acceptance tests that use `TestStep > ExpectError`. We can probably just let dependabot handle the upgrade when we release `terraform-plugin-docs` [v0.14.1](https://github.com/hashicorp/terraform-plugin-docs/milestone/9)
